[msb-compat] Handler from org.bson types to `Map[K, V]` uses view

Hi.

I am using the official mongo scala driver with RM's macros and reader/writer support. I wrote a `Codec[BSONValue]` myself to parse org.bson types to RM's types but since I discovered the msb-compat module, thought that I could use this one instead.

Seemed straight forward; I ran into some problems, however.

First thing is that decoding `BsonValue -> BSONValue -> MyCaseClass` seems to be successful, but when trying to use a property of the class of type `Map[String, M]`, I see an exception:

```scala
$ val codec: Codec[X] = BSONHandler[X]
$ val model = codec.decode(...) // OK, as long as you don't print it
...
$ println(model) // ->
// reactivemongo.api.bson.exceptions.HandlerException: Fails to handle default: BSONString != BSONDocument
// Caused by: reactivemongo.api.bson.exceptions.TypeDoesNotMatchException: BSONString != BSONDocument
```
That is caused by using the `mapValues` method from Utils, which calls `Map.mapValues` which just creates a view, and does not yet convert the values.

https://github.com/ReactiveMongo/ReactiveMongo-BSON/blob/c46185768be7f09e55af5ae0686e2f936210c081/api/src/main/scala/DefaultBSONHandlers.scala#L301-L303

I would expect the decode to fail right away if it can't parse the input.

--- a/api/src/main/scala-2.13-/TypesCompat.scala
+++ b/api/src/main/scala-2.13-/TypesCompat.scala
@@ -10,5 +10,7 @@ private[bson] trait Utils {
 
   @inline private[bson] def toLazy[T](it: Traversable[T]) = it.toStream
 
-  @inline private[bson] def mapValues[K, V, U](m: Map[K, V])(f: V => U): Map[K, U] = m.mapValues(f)
+  @inline private[bson] def mapValues[K, V, U](m: Map[K, V])(f: V => U): Map[K, U] = m.transform {
+    case (_, v) => f(v)
+  }
 }

--- a/api/src/test/scala/HandlerSpec.scala
+++ b/api/src/test/scala/HandlerSpec.scala
@@ -219,6 +219,15 @@ final class HandlerSpec extends org.specs2.mutable.Specification {
 
         handler.readTry(input) must beSuccessfulTry(expectedResult)
       }
+
+      "fail" in {
+        val input = BSONDocument(
+          "a" -> BSONDocument("label" -> "foo", "count" -> 10),
+          "b" -> BSONDocument("wrong" -> "foo2"))
+        val handler = implicitly[BSONReader[Map[String, Foo]]]
+
+        handler.readTry(input) must beFailedTry
+      }
     }
   }
 

